File: bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/widgets/List.java
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH. All rights
 * reserved. This program and the accompanying materials are made available
 * under the terms of the Eclipse Public License v1.0 which accompanies this
 * distribution, and is available at http://www.eclipse.org/legal/epl-v10.html
 * Contributors: Innoopract Informationssysteme GmbH - initial API and
 * implementation
 ******************************************************************************/
package org.eclipse.rap.rwt.widgets;

import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.events.SelectionEvent;
import org.eclipse.rap.rwt.events.SelectionListener;
import org.eclipse.rap.rwt.internal.widgets.IListAdapter;

// TODO [rh] H_SCROLL not yet implemented
public class List extends Scrollable {

  private final ListModel model;
  private int focusIndex = -1;
  private IListAdapter listAdapter;

  public List( final Composite parent, final int style ) {
    super( parent, checkStyle( style ) );
    model = new ListModel( ( style & RWT.SINGLE ) != 0 );
  }
  
  /////////////////////
  // Adaptable override
  
  public Object getAdapter( final Class adapter ) {
    Object result;
    if( adapter == IListAdapter.class ) {
      if( listAdapter == null ) {
        listAdapter = new IListAdapter() {
          public void setFocusIndex( final int focusIndex ) {
            List.this.setFocusIndex( focusIndex );
          }
        };
      }
      result = listAdapter;
    } else {
      result = super.getAdapter( adapter );
    }
    return result;
  }
  
  ///////////////////////////////
  // Methods to get/set selection

  public int getSelectionIndex() {
    return model.getSelectionIndex();
  }

  public int[] getSelectionIndices() {
    return model.getSelectionIndices();
  }
  
  public int getSelectionCount() {
    return model.getSelectionCount();
  }

  // TODO [rh] selection is not scrolled into view (see List.js)
  public void setSelection( final int selection ) {
    model.setSelection( selection );
    updateFocusIndexAfterSelectionChange();
  }

  public void setSelection( final int[] selection ) {
    model.setSelection( selection );
    updateFocusIndexAfterSelectionChange();
  }
  
  public void setSelection( final int start, final int end ) {
    model.setSelection( start, end );
    updateFocusIndexAfterSelectionChange();
  }
  
  public void setSelection( final String[] selection ) {
    model.setSelection( selection );
    updateFocusIndexAfterSelectionChange();
  }
  
  public void selectAll() {
    model.selectAll();
    updateFocusIndexAfterSelectionChange();
  }

  public void deselectAll() {
    model.deselectAll();
    updateFocusIndexAfterSelectionChange();
  }
  
  public int getFocusIndex() {
    return focusIndex;
  }

  ////////////////////////////////
  // Methods to maintain the items
  
  public void add( final String string ) {
    model.add( string );
    updateFocusIndexAfterItemChange();
  }

  public void add( final String string, final int index ) {
    model.add( string, index );
    updateFocusIndexAfterItemChange();
  }

  public void remove( final int index ) {
    model.remove( index );
    updateFocusIndexAfterItemChange();
  }

  public void remove( final int start, final int end ) {
    model.remove( start, end );
    updateFocusIndexAfterItemChange();
  }
  
  public void remove( final int[] indices ) {
    model.remove( indices );
    updateFocusIndexAfterItemChange();
  }

  public void remove( final String string ) {
    model.remove( string );
    updateFocusIndexAfterItemChange();
  }

  public void removeAll() {
    model.removeAll();
    updateFocusIndexAfterItemChange();
  }

  public void setItem( final int index, final String string ) {
    model.setItem( index, string );
  }

  public void setItems( final String[] items ) {
    model.setItems( items );
  }

  public String getItem( final int index ) {
    return model.getItem( index );
  }

  public int getItemCount() {
    return model.getItemCount();
  }

  public String[] getItems() {
    return model.getItems();
  }

  /////////////////////////////////////////
  // Listener registration/de-registration

  public void addSelectionListener( final SelectionListener listener ) {
    SelectionEvent.addListener( this, listener );
  }

  public void removeSelectionListener( final SelectionListener listener ) {
    SelectionEvent.removeListener( this, listener );
  }

  /////////////////////////////////
  // Helping methods for focusIndex 
  
  private void setFocusIndex( final int focusIndex ) {
    int count = model.getItemCount();
    if( focusIndex == -1 || ( focusIndex >= 0 && focusIndex < count ) ) {
      this.focusIndex = focusIndex;
    }
  }
  
  private void updateFocusIndexAfterSelectionChange() {
    if( model.getSelectionIndex() == -1 ) {
      focusIndex = 0;
    } else {
      focusIndex = model.getSelectionIndices()[ 0 ];
    }
  }

  private void updateFocusIndexAfterItemChange() {
    if( model.getItemCount() == 0 ) {
      focusIndex = -1;
    } else if( model.getSelectionIndex() == -1 ){
      focusIndex = model.getItemCount() - 1;
    }
  }
  
  //////////////////
  // Helping methods 
  
  private static int checkStyle( final int style ) {
    return checkBits( style, RWT.SINGLE, RWT.MULTI, 0, 0, 0, 0 );
  }
}

File: bundles/org.eclipse.rap.rwt/js/org/eclipse/rap/rwt/widgets/List.js
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

/**
 * This class extends qx.ui.form.List to make its API more SWT-like.
 */
qx.OO.defineClass( 
  "org.eclipse.rap.rwt.widgets.List", 
  qx.ui.form.List,
  function( multiSelection ) {
    qx.ui.form.List.call( this );
    // Should changeSelection events passed to the server-side? 
    // state == no, action == yes
    this._changeSelectionNotification = "state";
    this.setMarkLeadingItem( true );
    // qx.manager.selection.SelectionManager
    var manager = this.getManager();
    manager.setMultiSelection( multiSelection );
    manager.addEventListener( "changeSelection", this._onChangeSelection, this );
    manager.addEventListener( "changeLeadItem", this._onChangeLeadItem, this );
    this.addEventListener( "focus", this._onFocusIn, this );
    this.addEventListener( "blur", this._onFocusOut, this );
    this.addEventListener( "click", this._onClick, this );
    this.addEventListener( "dblclick", this._onDblClick, this );
    this.addEventListener( "changeEnabled", this._onChangeEnabled, this );
  } );

/** Sets the given aray of items. */
qx.Proto.setItems = function( items ) {
  // preserve selection and focused item
  var manager = this.getManager();
  var oldLeadItem = manager.getLeadItem();
  var oldAnchorItem = manager.getAnchorItem();
  var oldSelection = manager.getSelectedItems();
  // exchange/add/remove items
  var oldItems = this.getChildren();
  for( var i = 0; i < items.length; i++ ) {
    if( i < oldItems.length ) {
      oldItems[ i ].setLabel( items[ i ] );
    } else {
      // TODO [rh] optimize this: context menu should be handled by the List
      //      itself for all its ListItems
      var item = new qx.ui.form.ListItem( items[ i ] );
      // prevent items from being drawn outside the list
      item.setOverflow( qx.constant.Style.OVERFLOW_HIDDEN ); 
      item.setContextMenu( this.getContextMenu() );
      this.add( item );
    }
  }
  while( this.getChildrenLength() > items.length ) {
    this.removeAt( this.getChildrenLength() - 1 );
  }
  // restore previous selection and focusItem
  manager.setSelectedItems( oldSelection );
  manager.setLeadItem( oldLeadItem );
  if( manager.getMultiSelection() ) {
    manager.setAnchorItem( oldAnchorItem );
  }
}

/**
 * Sets the single selection for the List to the item specified by the given 
 * itemIndex (-1 to clear selection).
 */
qx.Proto.selectItem = function( itemIndex ) {
  if( itemIndex == -1 ) {
    this.getManager().deselectAll();
  } else {
    var item = this.getChildren()[ itemIndex ];
    this.getManager().setSelectedItem( item );  
    // TODO [rh] second parameter has no effect, figure out what it is for
    this.getManager().scrollItemIntoView( item, true );
  }
}

/**
 * Sets the multi selection for the List to the items specified by the given 
 * itemIndices array (empty array to clear selection).
 */
qx.Proto.selectItems = function( itemIndices ) {
  if( itemIndices.length == 0 ){
    this.getManager().deselectAll();
  } else {
    for( var i = 0; i < itemIndices.length; i++ ) {    
      var item = this.getChildren()[ itemIndices[ i ] ];
      this.getManager().setItemSelected( item, true );  
    }
  }
}

/**
 * Sets the focused item the List to the item specified by the given 
 * itemIndex (-1 for no focused item).
 */
qx.Proto.focusItem = function( itemIndex ) {
  if( itemIndex == -1 ) {
    this.getManager().setLeadItem( null );
  } else {
    var items = this.getManager().getItems();
    this.getManager().setLeadItem( items[ itemIndex ] );
  }
}

/**
 * Selects all item if the List is multi-select. Does nothing for single-
 * select Lists.
 */
qx.Proto.selectAll = function() {
  if( this.getManager().getMultiSelection() == true ) {
    this.getManager().selectAll();
  }
}

qx.Proto.setChangeSelectionNotification = function( value ) {
  this._changeSelectionNotification = value;
}

qx.Proto.setFont = function( value ) {
  var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
  var items = this.getChildren();
  for( var i = 0; i < items.length; i++ ) {
    wm.setFont( items[ i ], 
                value.getName(), 
                value.getSize(), 
                value.getBold(), 
                value.getItalic() );
  }
}

qx.Proto.dispose = function() {
  if( this.getDisposed() ) {
    return true;
  }
  var manager = this.getManager();
  manager.removeEventListener( "changeSelection", this._onChangeSelection, this );
  manager.removeEventListener( "changeLeadItem", this._onChangeLeadItem, this );
  this.removeEventListener( "focus", this._onFocusIn, this );
  this.removeEventListener( "blur", this._onFocusOut, this );
  this.removeEventListener( "click", this._onClick, this );
  this.removeEventListener( "dblclick", this._onDblClick, this );
  this.removeEventListener( "changeEnabled", this._onChangeEnabled, this );
  return qx.ui.form.List.prototype.dispose.call( this );
}

qx.Proto._onChangeSelection = function( evt ) {
  this._updateSelectedItemState();
  var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
  var id = wm.findIdByWidget( this );
  var req = org.eclipse.rap.rwt.Request.getInstance();
  req.addParameter( id + ".selection", this._getSelectionIndices() );
  if( this._changeSelectionNotification == "action" ) {
    this._suspendClicks();
    req.addEvent( "org.eclipse.rap.rwt.events.widgetSelected", id );
    req.send();
  }
}

qx.Proto._getSelectionIndices = function() {
  var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
  var id = wm.findIdByWidget( this );
  var selectionIndices = "";
  var selectedItems = this.getManager().getSelectedItems();
  for( var i = 0; i < selectedItems.length; i++ ) {
    var index = this.indexOf( selectedItems[ i ] );
    // TODO [rh] find out why sometimes index == -1, cannot be reproduced
    //      in standalone qooxdoo application
    if( index >= 0 ) {  
      if( selectionIndices != "" ) {
        selectionIndices += ",";
      }
      selectionIndices += String( index );
    }
  }
  return selectionIndices;
}

qx.Proto._onChangeLeadItem = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
    var id = wm.findIdByWidget( this );
    var req = org.eclipse.rap.rwt.Request.getInstance();
    var focusIndex = this.indexOf( this.getManager().getLeadItem() );
    req.addParameter( id + ".focusIndex", focusIndex );
  }
}

qx.Proto._onClick = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    if( this._changeSelectionNotification == "action" ) {
      if( !this._clicksSuspended ) {
        this._suspendClicks();
        var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
        var id = wm.findIdByWidget( this );
        var req = org.eclipse.rap.rwt.Request.getInstance();
        req.addEvent( "org.eclipse.rap.rwt.events.widgetSelected", id );
        req.send();
      }
    }
  }
}

qx.Proto._onDblClick = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    if( this._changeSelectionNotification == "action" ) {
      var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
      var id = wm.findIdByWidget( this );
      var req = org.eclipse.rap.rwt.Request.getInstance();
      req.addEvent( "org.eclipse.rap.rwt.events.widgetDefaultSelected", id );
      req.send();
    }
  }
}

/*
 * Suspends the processing of click events to avoid sending multiple
 * widgetSelected events to the server.
 */
qx.Proto._suspendClicks = function() {
  this._clicksSuspended = true;
  qx.client.Timer.once( this._enableClicks, this, 500 );
}

qx.Proto._enableClicks = function() {
  this._clicksSuspended = false;
}

/*
 * Passes enablement to list items
 */
qx.Proto._onChangeEnabled = function( evt ) {
  var newValue = evt.getData();
  var items = this.getChildren();
  for( var i = 0; i < items.length; i++ ) {
    var item = items[ i ];
    item.setEnabled( newValue );
  }
}

qx.Proto._onFocusIn = function( evt ) {
  this._updateSelectedItemState();
}

qx.Proto._onFocusOut = function( evt ) {
  this._updateSelectedItemState();
}

qx.Proto._updateSelectedItemState = function() {
  var selectedItems = this.getManager().getSelectedItems();
  for( var i = 0; i < selectedItems.length; i++ ) {
    if( this.getFocused() ) {
      selectedItems[ i ].addState( "focused" );
    } else {
      selectedItems[ i ].removeState( "focused" );
    }
  }
}
